lib/mllparty/connection_hub/client_monitor.ex

```diff
@@ -11,91 +10,129 @@ defmodule MLLParty.ConnectionHub.ClientMonitor do
-          end
-        end
-        # Reset down_at, down_notification_sent
-        {nil, false}
+        # Connection is up
+        handle_connection_up(endpoint, state)
       false ->
-        # Connection is down, check if it just went down
-        if state.down_at == nil do
-          Logger.info("#{endpoint} connection is down")
-          # Try to reconnect, only if the client is not already reconnecting
-          if not pending_reconnect do
-            Logger.info("Attempting to reconnect #{endpoint}")
-            # Client will periodically attempt to reconnect at it's configured interval
-            MLLP.Client.reconnect(client_pid)
-          end
-          # Return current time in seconds as down_at, false as down_notification_sent
-          {current_timestamp, false}
-        else
-          # The presence of down_at indicates, the connection has been down for a while
-          # Calculate the downtime duration
-          duration = current_timestamp - state.down_at
-          # If duration has been 60 seconds or more, send a slack notification (unless it was already sent)
-          sent =
-            if duration >= 60 and not state.down_notification_sent do
-              message = "#{endpoint} connection has been down for #{duration} seconds"
-              Logger.info(message)
-              SlackNotification.send_notification(message)
-              true
-            else
-              state.down_notification_sent
-            end
-          # Return current down_at, true as down_notification_sent
-          {state.down_at, sent}
-        end
+        # Connection is down
+        handle_connection_down(endpoint, connected, pending_reconnect, state)
     end
-    # Run again after 1 second
-    Process.send_after(self(), {:monitor, client_wrapper_pid}, 1000)
-    # Update the state
-    new_state = Map.merge(state, %{
-      down_at: down_at,
-      down_notification_sent: down_notification_sent
-    })
+    schedule_monitor()
+    new_state = Map.merge(state, new_state)
     {:noreply, new_state}
   end
 
-  # Get the current system timestamp in seconds
-  def timestamp do
-    {megasec, sec, _microsec} = :os.timestamp
-    megasec * 1000000 + sec
+  def handle_connection_up(endpoint, state) do
+    # Get current timestamp
+    now = DateTime.utc_now
+    # Connection is up, check if it was down previously
+    if state.down_at do
+      seconds = DateTime.diff(now, state.down_at)
+      message = "#{endpoint} connection is up. Last down duration: #{TimeUtils.format_seconds(seconds)}."
+      Logger.info(message)
+      # If a downtime notification was sent, follow up that it's connected again
+      if state.down_notification_sent do
+        SlackNotification.send_notification(message)
+      end
+    end
+    # Reset down_at, down_notification_sent, reconnect_attempted
+    %{down_at: nil, down_notification_sent: false, reconnect_attempted: false}
```

Comment:
Not a necessary change, but this is a spot where a slightly different design (pattern) can help make the code more predictable/durable. 

The change would be to change the function signature from:
```
handle_connection_up(endpoint, state) -> %{down_at: nil, down_notification_sent: false, reconnect_attempted: false}
```
to 
```
handle_connection_up(state) -> state
```

It's a small change, but possible since the `endpoint` arg is derivative of values contained in `state`, and the Map returned here is merged with `state` by the caller.

The change however makes the function a [monoid over state](https://fsharpforfunandprofit.com/posts/monoids-without-tears/), which makes it composable with any other functions with signature `func(state) -> state`, and also gives us the opportunity to add compiler-time type safety if we made state a struct (`State`). 

Functional programming languages don't really have as much as OO langs in the way of design patterns, but it's very common to make functions easily composable in this way because it minimizes the cognitive load of tracking what goes in and out, and allows programs to be constructed as pipelines.